### PR TITLE
Parameters sdram usage in bytes

### PR DESCRIPTION
**Pull request by Christian-B**
Replaces https://github.com/SpiNNakerManchester/sPyNNaker/pull/870 as helpful info from @pabogdan showed that was incorrect as it did not support multiple edges.

The DSG is now done at the Machine Level so is more acturate at better if new paritioining is used.

There is still other places at the App level but that is for future work.

Tested by: https://github.com/SpiNNakerManchester/sPyNNaker8/pull/448

### CODE REVIEW

**Review by alan-stokes — CHANGES_REQUESTED** [review]

**Comment by alan-stokes** [comment]
this ended up opening a can of worms in the skype chat when i asked what actually can be supported. this means i think this needs a bit of thinking to be as clear as possible. coz obviously it confuzzles us a lot. the fact that synapse dynamics setting causes a merge in pynn projection common AND this check makes it very confusing.

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural_common.py [hunk 1]
```diff
@@ -94,6 +99,10 @@ class SynapseDynamicsStructuralCommon(object):
     #: Default value for maximum fan-in per target layer neuron
     DEFAULT_S_MAX = 32
 
+    PAIR_ERROR = (
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural_common.py:102]: The error is correct.  You can only have one structural connection between each pair of Populations.  There can only be one structural type at the target, but that is checked elsewhere.

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural_common.py [hunk 2]
```diff
@@ -197,63 +202,84 @@ def write_parameters(
...
+        machine_edges = collections.defaultdict(list)
+        for machine_edge in machine_graph.get_edges_ending_at_vertex(
+                machine_vertex):
+            if isinstance(machine_edge, ProjectionMachineEdge):
+                for synapse_info in machine_edge.synapse_information:
+                    if isinstance(synapse_info.synapse_dynamics,
+                                  AbstractSynapseDynamicsStructural):
+                        app_edge = machine_edge.app_edge
+                        if app_edge in structural_edges:
+                            if (structural_edges[machine_edge] !=
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural_common.py:265]: YES!

[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural_common.py]: fixed

**Pull request merged**